### PR TITLE
Add `threadStatus` to IOSim and IOSimPOR

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This PR introduces an implementation of `threadStatus` for `IOSim`. It's currently a bit fast and loose on the "blocked reason" for threads (e.g. reading `MVar`s etc.) but should implement the rest of the functionality faithfully.

This fixes https://github.com/input-output-hk/io-sim/issues/12.

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
@coot do you perhaps have time to review this?

File: io-sim/test/Test/IOSim.hs
Comment (coot) [inline io-sim/test/Test/IOSim.hs:169]: Let's try to fit into 80 columns:
 
```suggestion
    [ testProperty "thread status finished (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_finished
    , testProperty "thread status finished (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_finished
    , testProperty "thread status running (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_running
    , testProperty "thread status running (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_running
    , testProperty "thread status blocked (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_blocked
    , testProperty "thread status blocked (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_blocked
    , testProperty "thread status blocked delay (IOSim)"
    $ wi1 $ runSimOrThrow prop_thread_status_blocked_delay
    , testProperty "thread status blocked delay (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_blocked_delay
    , testProperty "thread status died (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_died
    , testProperty "thread status died (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_died
    , testProperty "thread status died_own (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_died_own
    , testProperty "thread status died_own (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_died_own
    , testProperty "thread status yield (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_yield
    , testProperty "thread status yield (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_yield
    , testProperty "thread status mask (IOSim)"
    $ withMaxSuccess 1 $ runSimOrThrow prop_thread_status_mask
    , testProperty "thread status mask (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_mask
    , testProperty "thread status mask blocked (IOSim)"
    $ w1 $ runSimOrThrow prop_thread_status_mask_blocked
    , testProperty "thread status mask blocked (IO)"
    $ withMaxSuccess 1 $ ioProperty    prop_thread_status_mask_blocked
    ]

```

File: io-sim/test/Test/IOSim.hs
Comment (coot) [inline io-sim/test/Test/IOSim.hs:191]:
```suggestion
prop_two_threads_expect_ target prop =
  prop_two_threads_expect target (const $ yield) prop
```

File: io-sim/test/Test/IOSim.hs
Comment (coot) [inline io-sim/test/Test/IOSim.hs:193]: Could you also try to fit the other type signatures into 80 columns?
```suggestion
prop_thread_status_finished :: (MonadFork m, MonadDelay m, MonadThread m)
                            => m Property
```

Comment (coot) [reply io-sim/test/Test/IOSim.hs:193]: I need to link our [style guide](https://github.com/input-output-hk/ouroboros-network/blob/master/docs/StyleGuide.md) in this repo.

**Review by coot — CHANGES_REQUESTED** [review]
@MaximilianAlgehed thanks that looks good.  I see one problem with this implementation: it does not work well with our implementation of `MVar`'s which is based on `TVar`'s.  When a thread is blocked on an `MVar` we will thus report `ThreadBlocked BlockedOnSTM`. But it shouldn't be difficult to distinguished `MVar`s and `TVar`s.  We's need  to add a boolean flag to `TVar`: `tvarIsMVar :: Bool`.  The problem to solve is how to update it when we create an `MVar` in [`newEmptyMVarDefault`](https://github.com/input-output-hk/io-sim/blob/main/io-classes/src/Control/Monad/Class/MonadMVar.hs#L171-L172).  The simplest way would be to move that function to `io-sim`, where we have access to `TVar` internals.

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (MaximilianAlgehed) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:771]: This is probably just a left-over from experimenting that I missed when going over the diff.

File: io-sim/src/Control/Monad/IOSim/Internal.hs
Comment (MaximilianAlgehed) [reply io-sim/src/Control/Monad/IOSim/Internal.hs]: Yes, got rid of it.

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs
Comment (MaximilianAlgehed) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:102]: I realized when going over the code with John. Perhaps a welcome optimization would be to put the `FinishedReason` in here too then - to aovoid that look-up as well.

**Comment by MaximilianAlgehed** [comment]
@coot do you think we should try to address the `BlockReason` issues in this PR or do we postpone that for later?

**Comment by coot** [comment]
`BlockReason` could be addressed later; Please leave a comment which links to an issue.

**Comment by MaximilianAlgehed** [comment]
Issue mentioning `BlockReason` https://github.com/input-output-hk/io-sim/issues/21.

**Comment by MaximilianAlgehed** [comment]
@coot I think this should be about ready now.

**Review by dcoutts — COMMENTED** [review]
This looks ok to me.

I have to say I was not aware of the semantics of `threadStatus`. It looks like you believe that the difference between status died and finished is receiving an async exception.

The library docs merely say
```
ThreadFinished | the thread has finished
ThreadDied | the thread received an uncaught exception
```
without saying anything about async vs sync exceptions.

My own experiment is even less helpful, it suggests that `ThreadDied` isn't used at all!
```
import GHC.Conc
import Control.Exception

main = do
  --setUncaughtExceptionHandler throwIO

  tid <- forkIO $ do
           print "foo"
           threadDelay 10000000 --10sec
           print "bar"

  print =<< threadStatus tid
  throwTo tid (ErrorCall "oh noes")
  threadDelay 1000000
  print =<< threadStatus tid
```
and
```
 ghc-8.10.7 ThreadStatus.hs 
[1 of 1] Compiling Main             ( ThreadStatus.hs, ThreadStatus.o )
Linking ThreadStatus ...
[duncan@dunky io-sim]$ ./ThreadStatus 
ThreadRunning
ThreadStatus: oh noes
ThreadFinished
```
I thought it might be related to the default top-level exception handler for forkIO threads, but apparently not. If one uncomments the line `setUncaughtExceptionHandler throwIO` then the final result is instead
```
$ ./ThreadStatus 
ThreadRunning
ThreadRunning
```

I cannot get ghc to ever report `ThreadDied`.

If we don't need to track async vs sync exceptions for the cause here, then it'd simplify the patch quite a bit. If it were just threads terminating with an exception (any, sync or async) vs terminating via `return` then we'd not have to track so much in this patch.

File: io-sim/test/Main.hs
Comment (MaximilianAlgehed) [inline io-sim/test/Main.hs:15]: Whoops, I think we turned these off because they were flaky but not for any reason that we could see related to anything we did. @dcoutts any clue?

**Comment by MaximilianAlgehed** [comment]
@dcoutts did you check the tests? On my machine the tests pass - indicating that our hypothesis about the difference between synch and asynch exceptions is true *at least some of the time*.  See specifically the test
[prop_thread_status_died](https://github.com/input-output-hk/io-sim/pull/19/files#diff-843938f4a3c358e946651fcd9a6166fdb1cb567815fb8dfdbeae86f3a41c66cdR254-R263). Could you check out the tests and run them on your machine to see if they match?

**Comment by MaximilianAlgehed** [comment]
And to be clear, we were *very surprised* by the semantics we observed!

**Comment by MaximilianAlgehed** [comment]
@coot or @dcoutts do you mind running the CI here so we can get things to turn green?

**Review by coot — APPROVED** [review]
LGTM

**Comment by coot** [comment]
I don't see how to make CI run, there's no box which [github docs](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks) mentions :disappointed:.

**Comment by MaximilianAlgehed** [comment]
@dcoutts could you perhaps push the right buttons on this so we can get it merged?

**Comment by coot** [comment]
> @dcoutts could you perhaps push the right buttons on this so we can get it merged?

I asked similar question today.

Could you rebase the branch on top of `origin/main`: even if we run the workflows now they will fail without rebase (I checked that locally).

**Comment by coot** [comment]
@MaximilianAlgehed I pushed PR #27 with your changes (rebased), unfortunately two tests failed:
```
       thread status died_own (IO):             FAIL
        *** Failed! Falsified (after 1 test):
        ThreadFinished /= ThreadBlocked BlockedOnMVar
        Use --quickcheck-replay=392463 to reproduce.
        Use -p '/thread status died_own (IO)/' to rerun this test only.

       thread status mask (IO):                 FAIL
        *** Failed! Falsified (after 1 test):
        ThreadFinished /= ThreadBlocked BlockedOnMVar
        Use --quickcheck-replay=499493 to reproduce.
        Use -p '/thread status mask (IO)/' to rerun this test only.

```

**Comment by MaximilianAlgehed** [comment]
@coot that's a bit worrying - they pass locally. Are the tests run in parallel in CI? That might affect scheduling (these tests are understandably a bit sensitive to scheduling).

**Comment by coot** [comment]
The problem is only on Windows, where I can reproduce it locally too.

**Comment by coot** [comment]
Since `threadStatus` is not well specified, I think it's ok if we just disable these tests on Windows.

**Comment by MaximilianAlgehed** [comment]
@coot That's fine with me. I'll make an issue that mentions this so I have something to come back to after the next spate of deadlines.

**Comment by MaximilianAlgehed** [comment]
@coot I've made an issue here https://github.com/input-output-hk/io-sim/issues/28 and it's weirder than I thought! I finally got the time to look carefully at this and as I say in the issue the behaviour on Windows is probably just plain wrong! We don't do any operations that should put the thread that we ask `threadStatus` of in that state at any point in the execution!

**Comment by coot** [comment]
Thanks, I disabled the tests in the other PR.  Once CI will finish I'll merge it.

**Comment by coot** [comment]
Merged via #27.

**Comment by dcoutts** [comment]
I've filed a GHC issue: https://gitlab.haskell.org/ghc/ghc/-/issues/22279

**Comment by coot** [comment]
@dcoutts I created an issue to [track this](https://github.com/input-output-hk/io-sim/issues/38), let's close this PR as it's already merged via #27.

**Pull request closed without merging**
